File: service.controller.infrared/domain/OnkyoHTR380.js
const { Device } = require("../../libraries/javascript/device");

const irName = "ONKYO_HT_R380";

const inputOptions = [
  { value: "BD_DVD", text: "MacBook" },
  // { value: "VCR_DVR", text: "VCR/DVD" },
  { value: "CBL_SAT", text: "Roku" },
  { value: "GAME", text: "Other" },
  // { value: "AUX", text: "Auxiliary" },
  // { value: "TUNER", text: "Radio" },
  // { value: "TV_CD", text: "TV/CD" },
  // { value: "PORT", text: "PORT" }
];

const keys = {
  power: "KEY_POWER",
  BD_DVD: "KEY_DVD",
  VCR_DVR: "KEY_VCR",
  CBL_SAT: "KEY_SAT",
  GAME: "BTN_GAMEPAD",
  AUX: "KEY_AUX",
  TUNER: "KEY_TUNER",
  TV_CD: "KEY_TV",
  PORT: "KEY_TV2",
  volumeUp: "KEY_VOLUMEUP",
  volumeDown: "KEY_VOLUMEDOWN",
  mute: "KEY_MUTE"
};

class OnkyoHTR380 extends Device {
  constructor(config) {
    super(config);

    this.state = {
      power: { type: "bool" }
    };
  }

  // Perform validation and sanitization of the input
  conform(original) {
    let c = {};

    if ("power" in original) {
      c.power = Boolean(original.power);
    }

    if ("volume" in original) {
      if (!Number.isInteger(original.volume)) {
        throw new Error(`Volume must be an integer`);
      }

      // Volume is relative because we don't know what it's currently set to
      // but check that the number isn't ridiculous.
      if (Math.abs(original.volume) > 10) {
        throw new Error("Relative volume change must be [-10, 10]");
      }

      c.volume = original.volume;
    }

    if ("input" in original) {
      // Validate the input name
      if (!inputOptions.find(option => option.value === original.input)) {
        let options = inputOptions.map(option => option.value);
        throw new Error(
          `Invalid input '${original.input}'; must be one of: ${options}`
        );
      }

      // Make sure the key map is defined
      if (!(original.input in keys)) {
        throw new Error(`Key not defined for '${original.input}'`);
      }

      c.input = original.input;
    }

    if ("mute" in original && original.mute === true) {
      c.mute = true;
    }

    return c;
  }

  generateInstructions(state) {
    let instructions = [];

    if ("power" in state && state.power !== this.state.power.value) {
      instructions.push(irName, keys.power);

      // Give the AV receiver plenty of time to get its affairs in order
      instructions.push("wait", 5000);
    }

    if ("volume" in state) {
      const key = state.volume > 0 ? keys.volumeUp : keys.volumeDown;

      // Send the key n + 1 times because the key needs to be
      // pressed to activate the volume control before it changes
      for (let i = 0; i <= Math.abs(state.volume); i++) {
        instructions.push(irName, key);
        instructions.push("wait", 200);
      }

      // Wait until the volume control goes away
      // before attempting anything else
      instructions.push("wait", 3000);
    }

    if ("input" in state) {
      instructions.push(irName, keys[state.input]);
      instructions.push("wait", 3000);
    }

    if ("mute" in state) {
      instructions.push(irName, keys.mute);
      instructions.push("wait", 3000);
    }

    return instructions;
  }
}

exports = module.exports = OnkyoHTR380;

File: services/infrared/domain/onkyo_htr380.go
package domain

import (
	"context"
	"math"

	"github.com/jakewright/home-automation/libraries/go/device"
	devicedef "github.com/jakewright/home-automation/libraries/go/device/def"
	"github.com/jakewright/home-automation/libraries/go/oops"
	deviceregistrydef "github.com/jakewright/home-automation/services/device-registry/def"
	"github.com/jakewright/home-automation/services/infrared/ir"
)

const (
	onkyoHTR380KeyPower      = "KEY_POWER"
	onkyoHTR380KeyBDDVD      = "KEY_DVD"
	onkyoHTR380KeyVCRDVD     = "KEY_VCR"
	onkyoHTR380KeyCBLSAT     = "KEY_SAT"
	onkyoHTR380KeyGAME       = "BTN_GAMEPAD"
	onkyoHTR380KeyAUX        = "KEY_AUX"
	onkyoHTR380KeyTUNER      = "KEY_TUNER"
	onkyoHTR380KeyTVCD       = "KEY_TV"
	onkyoHTR380KeyPORT       = "KEY_TV2"
	onkyoHTR380KeyVolumeUp   = "KEY_VOLUMEUP"
	onkyoHTR380KeyVolumeDown = "KEY_VOLUMEDOWN"
	onkyoHTR380KeyMute       = "KEY_MUTE"

	onkyoHTR380InputBDDVD  = "BD_DVD"
	onkyoHTR380InputVCRDVD = "VCR_DVD"
	onkyoHTR380InputCBLSAT = "CBL_SAT"
	onkyoHTR380InputGAME   = "GAME"
	onkyoHTR380InputAUX    = "AUX"
	onkyoHTR380InputTUNER  = "TUNER"
	onkyoHTR380InputTVCD   = "TV_CD"
	onkyoHTR380InputPORT   = "PORT"
)

var onkyoHTR380InputKeys = map[string]string{
	onkyoHTR380InputBDDVD:  onkyoHTR380KeyBDDVD,
	onkyoHTR380InputVCRDVD: onkyoHTR380KeyVCRDVD,
	onkyoHTR380InputCBLSAT: onkyoHTR380KeyCBLSAT,
	onkyoHTR380InputGAME:   onkyoHTR380KeyGAME,
	onkyoHTR380InputAUX:    onkyoHTR380KeyAUX,
	onkyoHTR380InputTUNER:  onkyoHTR380KeyTUNER,
	onkyoHTR380InputTVCD:   onkyoHTR380KeyTVCD,
	onkyoHTR380InputPORT:   onkyoHTR380KeyPORT,
}

type OnkyoHTR380 struct {
	*deviceregistrydef.DeviceHeader
	power bool
}

func (d *OnkyoHTR380) ID() string {
	return d.Id
}

func (d *OnkyoHTR380) key(key string) ir.Instruction {
	return ir.Key("ONKYO_HT_R380", key)
}

func (d *OnkyoHTR380) LoadState(ctx context.Context) error {
	state, err := device.LoadProvidedState(ctx, d.Id, d.StateProviders)
	if err != nil {
		return oops.WithMessage(err, "failed to load provided state for device %q", d.Id)
	}

	if power, ok := state["power"].(bool); !ok {
		return oops.InternalService("state provider didn't provide power state for device %q", d.Id)
	} else {
		d.power = power
	}

	return nil
}

func (d *OnkyoHTR380) InstructionsFromState(state map[string]interface{}) ([]ir.Instruction, error) {
	if err := device.ValidateState(state, d.ToDef()); err != nil {
		return nil, err
	}

	var instructions []ir.Instruction

	if power, ok := state["power"].(bool); ok && power != d.power {
		instructions = append(instructions, d.key(onkyoHTR380KeyPower))

		// Give the AV receiver plenty of time to turn on
		instructions = append(instructions, ir.Wait(5000))
	}

	return instructions, nil
}

func (d *OnkyoHTR380) InstructionFromCommand(command string, args map[string]interface{}) ([]ir.Instruction, error) {
	if err := device.ValidateCommand(command, args, d.ToDef().Commands); err != nil {
		return nil, err
	}

	var instructions []ir.Instruction

	switch command {
	case "volume":
		delta, ok := args["delta"].(float64)
		if !ok {
			return nil, oops.BadRequest("invalid volume argument 'delta': %v", args["delta"])
		}

		key := onkyoHTR380KeyVolumeUp
		if delta < 0 {
			key = onkyoHTR380KeyVolumeDown
		}

		// Send the key n + 1 times because the key needs to be
		// pressed to activate the volume control before it changes
		for i := 0.0; i <= math.Abs(delta); i++ {
			instructions = append(instructions, d.key(key), ir.Wait(200))
		}

	case "mute":
		instructions = append(instructions, d.key(onkyoHTR380KeyMute), ir.Wait(2000))

	case "input":
		input, ok := args["input"].(string)
		if !ok {
			return nil, oops.BadRequest("invalid input argument 'input': %v", args["input"])
		}

		key, ok := onkyoHTR380InputKeys[input]
		if !ok {
			return nil, oops.InternalService("failed to find key for input option %q", input)
		}

		instructions = append(instructions, d.key(key), ir.Wait(2000))
	}

	return instructions, nil
}

//func (d *OnkyoHTR380) SetProperties(state map[string]interface{}) (bool, error) {
//	if err := device.ValidateState(state, d.ToDef().State); err != nil {
//		return false, err
//	}
//
//	changed := false
//
//	power, ok := state["power"].(bool)
//	if ok {
//		if d.power != power {
//			changed = true
//		}
//
//		d.power = power
//	}
//
//	return changed, nil
//}

func (d *OnkyoHTR380) ToDef() *devicedef.Device {
	return &devicedef.Device{
		Id:             d.ID(),
		Name:           d.Name,
		Type:           d.Type,
		Kind:           d.Kind,
		ControllerName: d.ControllerName,
		Attributes:     d.Attributes,
		StateProviders: d.StateProviders,
		State: map[string]*devicedef.Property{
			"power": device.BoolProperty(d.power),
		},
		Commands: map[string]*devicedef.Command{
			"volume": {
				Args: map[string]*devicedef.Arg{
					"delta": device.IntArg(-10, 10, true),
				},
			},
			"mute": {},
			"input": {
				Args: map[string]*devicedef.Arg{
					"input": device.StringArgWithOptions([]*devicedef.Option{
						{onkyoHTR380InputBDDVD, "Laptop"},
						{onkyoHTR380InputCBLSAT, "Roku"},
						{onkyoHTR380InputGAME, "Game"},
					}, true),
				},
			},
		},
	}
}

func (d *OnkyoHTR380) Copy() Device {
	return &OnkyoHTR380{
		DeviceHeader: d.DeviceHeader,
		power:        d.power,
	}
}
